dot/parachain/provisioner/prioritized_disputes_selection.go

```diff
@@ -15,10 +17,243 @@ import (
 	"github.com/ChainSafe/gossamer/lib/primitives"
 )
 
+const (
+	MaxDisputeVotesForwardedToRuntime = 200_000
+	VotesSelectionBatchSize           = 1_100
+)
+
 type BlockState interface {
 	GetRuntime(blockHash common.Hash) (instance parachain.RuntimeInstance, err error)
 }
 
+// Implements the `select_disputes` function which selects dispute votes which should
+// be sent to the Runtime.
+//
+// Provisioner fetches all disputes from `dispute-coordinator` and separates them in multiple
+// partitions. Please refer to `struct PartitionedDisputes` for details.
+//
+// Besides the prioritisation described above the votes in each partition are filtered too.
+// Provisioner fetches all onchain votes and filters them out from all partitions. As a result the
+// Runtime receives only fresh votes
+func SelectDisputes(
+	overseerChan chan<- any,
+	blockState BlockState,
+	leaf *parachaintypes.ActivatedLeaf,
+	maxDisputesVotes int,
+	voteSelectionBatchSize int,
+) parachaintypes.MultiDisputeStatementSet {
+	onchain, err := getOnchainDisputes(blockState, leaf.Hash)
+	if err != nil {
+		logger.Warnf("getting on chain disputes: %s", err.Error())
+		// Here we log the error and continue with an empty onchain set.
+		onchain = make(map[parachaintypes.DisputeKey]parachaintypes.DisputeState)
+	}
+
+	recent, err := requestDisputes(overseerChan)
+	if err != nil {
+		recent = []disputemessages.RecentDispute{}
+	}
+
+	// Filter out unconfirmed disputes except if already known onchain.
+	var filteredRecent []disputemessages.RecentDispute
+	for _, d := range recent {
+		if d.DisputeStatus.IsConfirmedConcluded() ||
+			containsOnchain(onchain, d.SessionIndex, d.CandidateHash) {
+			filteredRecent = append(filteredRecent, d)
+		}
+	}
+
+	partitioned := partitionRecentDisputes(filteredRecent, onchain)
+	voteResults, err := voteSelection(overseerChan, partitioned, onchain,
+		maxDisputesVotes, voteSelectionBatchSize)
+	if err != nil {
+		voteResults = []voteSelectionResult{}
```

Comment:
. when `voteSelection()` returns err, the logic here continues with empty `voteResults`; however,  it looks like `voteSelection()` only returns error when fails to `requestVotes()`, it will panic for other cases. In a general looking, it seems the big chunk of for loop logic in `voteSelection()` is not mandatory. Not sure if this is the expected logic

Reply:
if we sucessfully retrieve the votes from the `disputes-subsystem` then the for loop is mandatory to filter and transform all those collected votes.